### PR TITLE
JIT: Eliminate SumToSize by using Optional Lists

### PR DESCRIPTION
**Pull request by t-vi**
This PR is a eliminates unneeded grad_sum_to_size and in particular speeds up the LSTM backward by allowing better fusion.

It consists of two parts:
- In AutoDiff, record broadcasting sizes only if the broadcast output size is different from the input size, otherwise record None.
- The specialization of Optional arguments (#18407) allows us to then eliminate ` _grad_sum_to_size(t, None)` in the peephole optimization   step.

Thus, in the LSTM case, no SumToSize remain in the crucial fusion group. The trick here is that we can specialize on the runtime information from the forward.

I'm testing that different broadcasting situations lead to different graphs.

I didn't move all symbolic_script _grad_sum_to_size to the new logic, but it might be better to do this incrementally, anyway.

### CODE REVIEW

**Comment by ngimel** [comment]
It would be nice to add a test, e.g. run scripted function for some inputs, then test that `.graph_for` for inputs with the different broadcasting pattern errors out.

**Comment by apaszke** [comment]
It would also be great if we could get some perf numbers to ensure that we don't slow down forward.

**Comment by t-vi** [comment]
Unfortunately we do seem to slow down the forward significantly (4%).

**Comment by apaszke** [comment]
Ok, that should be fixable, because it's not like we're doing significantly more work.

**Comment by t-vi** [comment]
@ngimel pointed out a problem with this: The forward pass seems to instantiate intermediate results where previously we didn't.

**Comment by t-vi** [comment]
Seems like this needs a rather thorough rebase, I'll try to look into it soonish.

**Comment by t-vi** [comment]
Numbers for the default benchmark config for me:

This

name | avg_fwd | std_fwd | avg_bwd | std_bwd
-|-|-|-|-
jit | 8.667 | 0.1126 | 12.91 | 0.2302
jit_premul | 7.204 | 0.0088 | 10 | 0.07453
cudnn | 6.182 | 0.04415 | 7.533 | 0.09015

Master

 | name | avg_fwd | std_fwd | avg_bwd | std_bwd
-|-|-|-|-
 | jit | 8.68 | 0.05878 | 14.88 | 0.1656
 | jit_premul | 7.239 | 0.008265 | 12.13 | 0.03549
 | cudnn | 6.197 | 0.03018 | 7.536 | 0.07714

**Comment by t-vi** [comment]
So unless wild things happen in the CI (and aren't already happening in master), I think this might be good.
:rocket: 

Thank you for all your help on this and the related patches.

File: torch/csrc/jit/autodiff.cpp [hunk 1]
```diff
@@ -213,11 +213,15 @@ class GradientHelper {
...
-  SymbolicVariable gradSumToSizeOf(SymbolicVariable v, Symbol input_name) {
+  SymbolicVariable gradSumToSizeOf(
+      SymbolicVariable v,
+      Symbol input_name,
+      SymbolicVariable fw_output) {
     Value* size;
     {
-      WithInsertPoint insert_guard{node};
-      size = SymbolicVariable(node->namedInput(input_name)).size();
+      WithInsertPoint insert_guard{node->next()};
```
Comment (wanchaol) [inline torch/csrc/jit/autodiff.cpp]: why does this guard changed from node to node->next()?

File: torch/csrc/jit/passes/peephole.cpp [hunk 2]
```diff
@@ -157,11 +163,12 @@ void PeepholeOptimizeImpl(Block* block, bool addmm_fusion_enabled) {
       }
     } else if (
         node->matches(
-            "aten::_grad_sum_to_size(Tensor(a) self, int[] size) -> Tensor(a)")) {
+            "aten::_grad_sum_to_size(Tensor(a) self, int[]? size) -> Tensor(a)")) {
```
Comment (wanchaol) [inline torch/csrc/jit/passes/peephole.cpp]: can you just merge above `_grad_sum_to_size` case with this one here?

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 3]
```diff
@@ -861,6 +862,8 @@ struct GraphFuser {
     // The output of producer_for_chunk_node could have been used in some
     // aten::size operators, so we need to clean those up as well (we simply
     // broadcast all its tensor inputs).
+    // We need to insert these earlier in the graph.
+    WithInsertPoint guard2(producer_for_chunk_node);
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/csrc/jit/passes/graph_fuser.cpp]: can you make this into the code comment so that we could refer to in the future?

**Review by wanchaol — APPROVED** [review]
looks nice! I think it's good to go, only have some nits :)

File: torch/csrc/jit/autodiff.cpp (hunk 1 above)
Comment (t-vi) [reply torch/csrc/jit/autodiff.cpp]: I added this as a comment, too: We want to use the output of node.

**Comment by wanchaol** [comment]
hmm it seems like the test failures are legit..

**Comment by t-vi** [comment]
So the failure `pytorch_linux_xenial_cuda9_cudnn7_py2_test` is in "my test". I'm trying to output the graph to see what's going on.
I don't have as many ideas about the other failures.

**Pull request closed without merging**